app/src/main/java/nl/eduvpn/app/viewmodel/ConnectionStatusViewModel.kt

```diff
@@ -93,12 +103,24 @@ class ConnectionStatusViewModel @Inject constructor(
     override fun onResume() {
         super.onResume()
         runUpdateCertExpiryEverySecond()
+        cancelAllExpiryNotifications()
     }
 
     fun onPause() {
+        planExpiryNotification()
         stopUpdateCertExpiry()
     }
 
+    private fun planExpiryNotification() {
+        if (certExpiryTime != null && vpnService.status != VPNService.VPNStatus.DISCONNECTED) {
+            alarmManager.setExact(AlarmManager.RTC, certExpiryTime, pendingIntent)
```

Comment:
Do we have a permission for this?
> Starting with Build.VERSION_CODES#S, apps require the SCHEDULE_EXACT_ALARM permission to use this API.
[From here](https://developer.android.com/reference/android/app/AlarmManager#setExact(int,%20long,%20android.app.PendingIntent))

Reply:
I have added the SCHEDULE_EXACT_ALARM permission in https://github.com/eduvpn/android/pull/333/commits/4122ad38bc07ff1e8e608faad54259b57463898f.

> feel free to merge

I do not have the permissions necessary to merge.

Reply:
Does the user need to approve the application for this new permission?

Reply:
This permission is only available Android >= 12, what effect does it have on Android < 12?

Reply:
> This permission is only available Android >= 12, what effect does it have on Android < 12?

Adding the permission should not change any behavior because our current target sdk is 29 (Android 10) and the permission is only used on Android >= 12. This will only change app behavior when we update our target sdk to 31 (Android S).

> Does the user need to approve the application for this new permission?

Surprisingly, the [user has to give the app a special permission on Android >=12](https://developer.android.com/about/versions/12/behavior-changes-12#exact-alarm-acceptable-use-cases) to use setExact. To save battery I decided to use  `AlarmManager.RTC` which does not wake the device up and should thus not affect battery usage. This is apparently not enough for Android 12 and instead of asking the user permission for a precise notification I think we should set a window of 5 minutes before expiry in which the notification can be send by the system.

Reply:
Problem with this is that when the user gets a notification 5 minutes before expiry, the user still has to wait for 5 minutes before he can request a new certificate. To fix this we should add a possibility to request a new certificate even if the current one has not yet expired.

Reply:
> Problem with this is that when the user gets a notification 5 minutes before expiry, the user still has to wait for 5 minutes before he can request a new certificate. To fix this we should add a possibility to request a new certificate even if the current one has not yet expired.

Why does the user still have to wait 5 minutes?

Reply:
The app will use the certificate until it has expired.
When sending the notification 5 minutes before expiry, the following will happen:
1. User gets notification that certificate will expire in 5 minutes
2. User opens app to get new certificate
3. User disconnects VPN
4. User reconnects
5. Certificate will expire in 4 minutes
6. User is annoyed for 4 minutes
7. App disconnects
8. User reconnects

Reply:
Right. That is why the OAuth authorization needs to be deleted as well before requesting the new certificate so the user logs in again. That is not currently implemented?

Reply:
I guess the "Renew" behavior is not specified anywhere...